tests/Fixer/FunctionNotation/NullableTypeDeclarationForDefaultNullValueFixerTest.php

```diff
@@ -166,6 +166,13 @@ public static function provideFixCases(): iterable
             '<?php function foo(? /*comment*/ string $param = null) {}',
             ['use_nullable_type_declaration' => false],
         ];
+
+        yield 'do not create implicitly nullable parameter types' => [
+            // @see https://wiki.php.net/rfc/deprecate-implicitly-nullable-types
+            '<?php function foo(?int $a = null, string $b) {}',
+            null,
```

Comment:
I think the test case I would expeect is one that turn 1 into 2:
1. `'<?php function foo(int $a = null, string $b) {}',`
2. `'<?php function foo(?int $a, string $b) {}',`

Reply:
Removing ` = null` is not the responsibility of `nullable_type_declaration_for_default_null_value`, but `no_unreachable_default_argument_value`, and using both is needed to achieve that transformation, which is happening already, without this fix: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.49.0/tests/Fixtures/Integration/priority/nullable_type_declaration_for_default_null_value%2Cno_unreachable_default_argument_value.test

Reply:
@kubawerlos This fix prevents the fixer to create implicit nullable param when there are required params, which is somehow confusing - it does not fix the main issue (optional param being before required param) but at the same time it deviates from the behaviour under `use_nullable_type_declaration=false` as it does not do what's the sole purpose of this fixer 🤔.

I also don't get what's the difference between current [example 2](https://cs.symfony.com/doc/rules/function_notation/nullable_type_declaration_for_default_null_value.html#example-2) and the test case in this PR.

```diff
-- Original
+++ New
 <?php
-function sample(?string $str = null)
+function sample(string $str = null)
 {}
```

I mean, I know that the difference is that one has only one parameter, and the other has two, but when it comes to [the RFC](https://wiki.php.net/rfc/deprecate-implicitly-nullable-types), one prevents deprecated syntax, while the other not. I am not sure if it should work this way. Just thinking out loud 🙂.

> I think the test case I would expeect is one that turns 1 into 2:
> 
> 1. `'<?php function foo(int $a = null, string $b) {}',`
> 2. `'<?php function foo(?int $a, string $b) {}',`

I believe this is out of the scope of _this particular fixer in its current state_, which is about handling nullable type (`?` or `|null`) for params with default `null` value. It does _not_ touch the default value, and even if adding such feature is technically possible, it would make this rule [risky](https://3v4l.org/ZfqBS) (compared to [this](https://3v4l.org/sk8O3)). I believe in this case you should just use `nullable_type_declaration_for_default_null_value` with `use_nullable_type_declaration=true` + there should be a separate, risky fixer which would remove default value for params not being _after_ required params.

Reply:
From the end user pov, the currnet behavior is the risky one, because it turns a non-deprecated syntax `(int $a = null, $b)` into a deprecated one `(?int $a = null, $b)`.

Reply:
> After more thinking about this, I now think all the whole ['use_nullable_type_declaration' => false] strategy is about creating implicitly nullable parameter types, which is going to be deprecated in PHP 8.4.
> 
> Maybe the right direction is to deprecate that strategy in this fixer?

I wanted to write it before and I forgot after writing all that thoughts 😆. Yeah, we should deprecate the `nullable_type_declaration_for_default_null_value` option and from next major version this fixer should only ensure explicit nullability.

Reply:
> From the end user pov, the current behavior is the risky one, because it turns
> a non-deprecated syntax:  `(int $a = null, $b)`
> into a deprecated one:  `(?int $a = null, $b)`.

Only considering the RFC, which was created way after the rule 😉. Previously it was only a matter of preference.

Anyway, it only shows we should deprecate the config option and allow only enforcing explicit nullability in the future.

Reply:
What about this fixer: https://cs.symfony.com/doc/rules/language_construct/nullable_type_declaration.html ? Maybe this one should take care of `int $a = null` when `null` is not default value (as followed by param without default)?

Reply:
`nullable_type_declaration` also does not cover default values, only standardise the way how nullability is defined (`?` vs `|null`). IMHO if we want to fix `foo($a = null, $b)` (optional before required) we need a new, risky fixer.